independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java

```diff
@@ -323,9 +323,15 @@ private static ScopeInfo initStereotypeScope(List<StereotypeInfo> stereotypes, A
         if (stereotypes.isEmpty()) {
             return null;
         }
-        final Set<ScopeInfo> stereotypeScopes = new HashSet<>();
-        for (StereotypeInfo stereotype : stereotypes) {
-            stereotypeScopes.add(stereotype.getDefaultScope());
+        // first go through all standard stereotypes
+        Set<ScopeInfo> stereotypeScopes = stereotypes.stream().filter(s -> !s.isAdditionalBeanDefiningAnnotation())
+                .map(stereotypeInfo -> stereotypeInfo.getDefaultScope())
+                .collect(Collectors.toSet());
+        // if the set is empty, try adding all stereotypes originating from additional bean def. ann.
+        if (stereotypeScopes.isEmpty()) {
+            stereotypeScopes.addAll(stereotypes.stream().filter(s -> s.isAdditionalBeanDefiningAnnotation())
```

Comment:
There's no need to `filter` the stream here, right?

Reply:
You either filter it or you attempt to re-add those that you already went over, which is harmless anyway.

Reply:
If it's empty then there are no non-bean defining annotations stereotypes, or?